[MRG+2] Refactor CategoricalEncoder into OneHotEncoder (with deprecated kwargs) and OrdinalEncoder

jorisvandenbossche opened the pull request:
Possibly fixes https://github.com/scikit-learn/scikit-learn/issues/10521

This does split the CategoricalEncoder into two separate classes for onehot and ordinal encoding, and then integrates the onehot encoding into the existing OneHotEncoder.

I also moved them to a separate file, so for reviewing actual changes might be better to not view the first commit.

I also did not yet introduce deprecation warnings for the old kwargs / attributes (or computed the new attributes in the old setting), but I infer based on the passed data to fit whether it was accepted before (and in this case we should raise a deprecation warning) and if not directly use the new behaviour. This 'inferring' of the behaviour can be overwritten with a newly added a `encoded_input=True/False` keyword.

The main drawback for new users of the OneHotEncoder is the 'pollution' with the old keywords and attributes in the docstring, repr of the object and tab completion.

jorisvandenbossche committed:
REF: move encoders to separate file

jorisvandenbossche committed:
Split CategoricalEncoder in OneHotEncoder and OrdinalEncoder

jorisvandenbossche committed:
Combine old and new OneHotEncoder in one class

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-bf48864f0eb289111c650bea55df1dfc16b47774)

**new alerts:**

* 1 for Explicit export is not defined

---

*Comment posted by [lgtm.com](https://lgtm.com)*

jorisvandenbossche committed:
Add main deprecation warning and deprecated attributes

jorisvandenbossche committed:
deprecate old keyword arguments

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-87b3b39b22c4d1edb24f5c20778304d12f9d9b6d)

**new alerts:**

* 1 for Explicit export is not defined

---

*Comment posted by [lgtm.com](https://lgtm.com)*

jorisvandenbossche committed:
make sure repr and get_params don't raise deprecation warnings

jorisvandenbossche committed:
remove remaining occurences of CategoricalEncoder in codebase

jorisvandenbossche committed:
fix doctest

jorisvandenbossche commented:
In the meantime, I added deprecation warnings for the old behaviour, keywords and attributes.
(still need to update deprecation messages and the docs)

jorisvandenbossche committed:
Don't show deprecated kwargs in class repr

jnothman commented on sklearn/base.py:
```diff
@@ -225,12 +225,27 @@ def get_params(self, deep=True):
             Parameter names mapped to their values.
         """
         out = dict()
-        for key in self._get_param_names():
```
This is getting a bit adventurous of you! Propose this separately. It's not exclusive to this change.

jorisvandenbossche replied:
It somehow is, in the sense that it is quite essential for this PR, as otherwise just showing the repr of the new OneHotEncoder shows deprecation warnings. 

Did it happen before that keyword arguments were deprecated? How was it dealt with then?

I can certainly do it in a separate PR, but then that PR would be a blocker for this one IMO (which is not necessarily a problem, so fine for me to do that).

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,808 @@
+        The categories of each feature determined during fitting
+        (in order corresponding with output of ``transform``).
+
+    Deprecated Attributes
```
I'm sure Numpydoc doesn't handle this.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,808 @@
+        will be all zeros. In the inverse transform, an unknown category
+        will be denoted as None.
+
+    Deprecated Parameters
```
I'm sure Numpydoc doesn't handle this. Just use `.. deprecated:: 0.20`, yeah?

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,808 @@
+        is set to 'ignore' and an unknown category is encountered during
+        transform, the resulting one-hot encoded columns for this feature
+        will be all zeros. In the inverse transform, an unknown category
+        will be denoted as None.
```
(Perhaps add a note that this can be used to handle missing values)

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,808 @@
+
+        encoded_input=False : categorical features that still need to be
+            encoded.
+        encoded_input=True : already integer encoded data, and the categories
```
"In the range `0...(n values - 1)`"

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,808 @@
+
+        The used categories can be found in the ``categories_`` attribute.
+
+    encoded_input : boolean
```
I think `ordinal_input` is better than `encoded_input`, especially seeing as we introduce an `OrdinalEncoder`...

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,808 @@
+    encoded_input : boolean
+        How to interpret the input data:
+
+        encoded_input=False : categorical features that still need to be
```
Don't use ` : descr`. The correct syntax for definition lists is:

```
encoded_input=False
    categorical features ...
```

I think the default should be 'auto', not None, with the default changing to False in two versions.

Do you think this option will be useful (for efficiency, for error handling, or for quality assurance) after the deprecation is finished? If not, should it just be called `legacy_mode`?

Does this setting affect the handling of `n_values`? Can  one use `encoded_input=True` and `categories` together?

This seems to mostly affect the handling of errors, in that even if a column isn't active in input, that value is not considered unknown....?

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,808 @@
+    OneHotEncoder(categories='auto', dtype=<... 'numpy.float64'>,
+           encoded_input=True, handle_unknown='error', sparse=True)
+    >>> enc.transform([[0, 1, 1]]).toarray()
+    array([[ 1.,  0.,  0.,  1.,  0.,  0.,  1.,  0.,  0.]])
```
Didn't this used to only show active features? Shouldn't this have columns for [x0=0, x0=1, x1=0, x0=1, x0=2, x2=0, x2=1, x2=3]? Here it has 9 columns, not 8.

jnothman replied:
Didn't this used to only show active features? Shouldn't this have columns for [x0=0, x0=1, x1=0, x0=1, x0=2, x2=0, x2=1, x2=3]? Here it has 9 columns, not 8.

jnothman reviewed (changes requested):
Clarifying the difference between encoded_input=False/True, and what will happen to it in two versions' time, still needs elucidation in the docs.

jorisvandenbossche commented:
Some quick answers

> Do you think this option will be useful (for efficiency, for error handling, or for quality assurance) after the deprecation is finished? If not, should it just be called `legacy_mode`?

See my (long) comment on the issue asking about this: https://github.com/scikit-learn/scikit-learn/issues/10521 
Short answer: I personally don't really know if it (after deprecation) is worth its own option. Would appreciate input there (as indeed the design would change a bit then)

> Does this setting affect the handling of n_values? Can one use encoded_input=True and categories together?

Good question, this is one of the things I didn't really think through yet. In principle, you can pass `categories`, but using the same format as for categorical input data does not make much sense (you would need to pass `categories=[[0, 1, 2, 3, 4], [0, 1, 2]]` for 5 and 3 categories, and we would need to check that it is consecutive range. 
But changing the way to specify `categories` depending on `encoded_input=True` (eg by giving the number of categories like `categories=[5, 3]` like the current `n_values`) is also not the nicest API design (but maybe the better solution ..)

> I'm sure Numpydoc doesn't handle this.

Yes I know, but wanted it to be very explicit now to see the consequences (for initial reviewing). I didn't bother yet fully updating the documentation as well.

jnothman commented:
I think we may have previously handled such warnings in get_params using a
check_warnings context, IIRC, but it resulted in dangerous race conditions
in parallel processing.

We usually don't bother raising warnings for constructor parameter access:
if the user doesn't have that parameter set at fit, they are unlikely to
try to access it otherwise in saved code. I know, it's a bit dodgy.

I personally think legacy_mode='auto' is the way to go, switching on the
basis of the data whether to use only the new params/attributes, or the
old. Most people don't care about the differences between the old and new
model. It can either default to False in v0.22 then disappear in v0.23 or
0.24, or False can become the *only* behaviour in v0.22 before it
disappears in v0.23 or 0.24.

jorisvandenbossche committed:
very messy first try to handle deprecations (infer legacy mode)

jorisvandenbossche commented:
Last commit added the logic to deal with all different case (to raise a warning or not, to use legacy mode or not). Still have to clean-up a lot, so don't review in detail, but could already take a look at `_handle_deprecations` method.

jorisvandenbossche committed:
Add categories='auto' as way to force new behaviour

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,850 @@
+
+    def _handle_deprecations(self, X):
+
+        if self.categories != 'auto':
```
Let's make `self.categories='legacy'` by default, so that the user can select `auto` explicitly.

jorisvandenbossche replied:
Yesterday I made a commit to make the default `categories=None` for the same purpose, but only pushed it now (the last commit). 
I personally find that cleaner, as `'legacy'` would not give the correct meaning if you are using string data.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,850 @@
+        self.dtype = dtype
+        self.handle_unknown = handle_unknown
+
+        if n_values is not None:
```
why don't you just use `self.n_values = n_values`? The warning is done there...

jorisvandenbossche replied:
> why don't you just use self.n_values = n_values? The warning is done there...

Because I also want to deprecated the access / writing of the attribute `n_values` on the class object.

jnothman replied:
Huh? But doing `self.n_values = n_values` here will call the setter and raise the warning, just as you have done.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,850 @@
+
+        if self._legacy_mode:
+            # TODO not with _transform_selected ??
+            self._fit_transform_old(X)
```
call it `_legacy_fit_transform`

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,850 @@
+                  ``X[:, i]``. Each feature value should be
+                  in ``range(n_values[i])``
+
+    categorical_features : "all" or array of indices or mask
```
I'm not sure users will be happy that we are deprecating this! But I suppose that in the deprecation notice, we'll be able to point to ColumnTransformer?

jorisvandenbossche replied:
regarding `categorical_features`: in principle I can keep this as this is not related to the inherent behaviour of how the encoding works in OneHotEncoder (legacy or new behaviour), but: it makes the implementation more complex, is not done in any other transformer, and indeed can be replaced with ColumnTransformer.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,850 @@
+                    self._legacy_mode = True
+
+        if self._deprecated_categorical_features != 'all':
+            self._legacy_mode = True
```
This isn't safe to do if `categories` has been set

jorisvandenbossche replied:
Yes, I know, but I don't want to implement this ability for the new behaviour. 
And, either you were using OneHotEncoder already, and then `categories` was not set and this is OK, or either you did already update your usage (eg setting `categories` instead of n_values) but then you can directly update for this deprecated keyword as well, or either you are new to this class and then you shouldn't use it.

What I can do is detect that `categories` is set by the *user* (and not internally set), and in that case just raise a plain error here instead of a warning.

jnothman replied:
Absolutely. Error if categorical_features is set and not legacy_mode

jnothman reviewed (commented):
I've not looked at fit/transform

jnothman reviewed (commented):
I've not looked at fit/transform

jorisvandenbossche committed:
Cleaning up deprecation warnings / tests

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into categorical-refactor

amueller commented:
Test failing? Is this waiting for reviews?

jorisvandenbossche commented:
This is waiting for your opinion about the general idea in https://github.com/scikit-learn/scikit-learn/issues/10521 (I would say, let's keep the general discussion there for now, would need to look back at this PR to know its actual status)

jorisvandenbossche committed:
temp

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into categorical-refactor

jorisvandenbossche committed:
clean-up

jorisvandenbossche commented:
@jnothman @amueller I looked at this again, did some small code clean-up, but this is actually ready for review 
(I only did not yet update the names in the docs)

amueller commented:
test failure still?

amueller commented on setup.cfg:
```diff
@@ -6,7 +6,7 @@ test = pytest
 addopts =
     --doctest-modules
-    --disable-pytest-warnings
+#    --disable-pytest-warnings
```
what's this?

jorisvandenbossche replied:
Sorry, I committed that temporarily because I need to see any warnings generated by the tests (to check I captured all deprecation warnings). The problem with this flag is that it hides any warning, not only the ones coming from pytest.

amueller replied:
wait and in master we're suppressing them?

jorisvandenbossche replied:
> wait and in master we're suppressing them?

yep, which is not good IMO, but we first need to fix the pytest warnings before enabling the warnings globally becomes useful (otherwise there are way too many pytest warnings to see the relevant ones). 
I think @rth is working on that

jorisvandenbossche replied:
Also in general, if we want to remove this hiding of the warnings, we should start catching all warnings that are knowingly raised in the tests (as I am doing in this PR), because looking at the travis log from this PR, that clearly does not happen at the moment (there are many deprecation/future/convergence warnings raised in the tests)

amueller replied:
I 100% agree with you and when I saw the travis output last week I was a bit shocked ;) Doing this has been one of my constant struggles in creating the releases.

ogrisel replied:
I opened #11252 to track fixing `DeprecationWarning`s and `FutureWarning`s in tests.

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+
+        Non-categorical features are always stacked to the right of the matrix.
+
+        .. deprecated:: 0.20
```
I'm somewhat tempted to not deprecate this and leave it in as a convenience. This is **so** common and would make sklearn so much more user friendly and not deprecating it won't kill us.
At least let's keep it until we have some other way to express #10603.

jnothman replied:
None strong.

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+    def __init__(self, n_values=None, categorical_features=None,
+                 categories=None, sparse=True, dtype=np.float64,
+                 handle_unknown='error'):
+        self._categories = categories
```
errr? why? that seems not good?

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+        self.handle_unknown = handle_unknown
+
+        if n_values is not None:
+            pass
```
All validation should be in ``fit``?

jorisvandenbossche replied:
Yes, I need to clean this up

amueller reviewed (commented):
I don't understand why you do the deprecation the way you do. Can you elaborate? I'll finish the review once I understand better what's happening.

jorisvandenbossche commented:
@amueller:

> I don't understand why you do the deprecation the way you do. Can you elaborate? I'll finish the review once I understand better what's happening.

For the deprecation, I wanted to do it "properly" (but of course different definitions of properly can exist :-)), so wanted the following:

- Deprecation warning when using deprecated keywords in init (they are currently raised during `fit`)
- Deprecation warnings when accessing deprecated attributes (`active_features_`, `feature_indices_`, `n_values_`).
  - For this I turned them into a property with getter/setter, the actual value is internally stored in an underscored version.
- Deprecation warnings for accessing / setting attributes that come from deprecated keywords (`n_values`, `categorical_features`). 
  - This is the part where Joel mentioned this might not be needed (https://github.com/scikit-learn/scikit-learn/pull/10523#issuecomment-359962473), but I would argue that if it is not too hard to do it, we should try to do it (as otherwise there will be some (probably minor) use cases where no deprecation warnings are raised appropriately)
  - For this, I also turned them into property with getter/setter.

The problem with this last one, is that we need to ensure that `get_params()` does not raise any deprecation warning. But since `get_params()` accesses the params as attributes on the class, it will give warnings. Therefore, I added a little bit of logic to `get_params()` to check if a param is deprecated (the new rule would be that in such a case it has an internal `_deprecated_[param_name]` version that does not raise a warning), and in that case use this internal attribute to ensure no warnings are raised from `get_params()`.

This is the added logic to `base.py`. Personally, I think the code is rather simple (it does not try to do any clever python trick as was tried before, but is just based on a naming convention that we can document for sklearn contributors), and therefore worth to add it to be able to properly do the deprecation of the keywords. 
The alternative is to not deprecate the attribute access/setting (there might be other ways solve it as well, though, but I don't directly see one that also ensures `get_params()` does not raise a warning).

There is one extra bit of code that I added to `base.py` which is to have the `__repr__` only show non-deprecated keywords. I think this would be nice for the OneHotEncoder, but this is rather separate from the rest of the code changes to `base.py`, so this I could easily remove.



- Additionally, I wanted to have a "clean" repr that does not include the deprecated keywords, to not

amueller commented:
``sklearn.preprocessing.data.OneHotEncoder`` was public and we're deleting it, right? We should really avoid public submodules. Let's deprecate ``.data`` but I don't think we can remove it like this.

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+                 handle_unknown='error'):
+        self._categories = categories
+        if categories is None:
+            self.categories = 'auto'
```
Opened #11117 I think this is not good.

jorisvandenbossche replied:
Hmm, do you have an alternative suggestion how to deal with this? 

I basically wanted to ensure that the `.categories` attribute (and the value you see in `get_params` or in the repr ) embodies the *actual* default ('auto'), and not the `None` whose only purpose is to check if the user passed the value for `categories` explicitly or not.

amueller replied:
Again, this is not how we have been doing things. It would be nice to have the "right" value in the ``__repr__`` and ``get_params``, but do we want to start this now? @GaelVaroquaux @jnothman  opinions? So far we did everything in ``fit`` and basically didn't care about repr and get_params being weird for deprecations.

jorisvandenbossche replied:
The main reason I find it important in this case is that I want to make the experience for *new* users of the OneHotEncoder (the ones that don't have the baggage of how it was before, and just want to start using the new functionality of finally being able to one-hot-encode string categories) as smooth as possible. And those don't need to wonder what this None means (although they will see it in the signature of course), and they don't need to see the deprecated keywords in the repr.

To be clear, I don't have problems with being pointed out how things are typically done in sklearn (I don't yet have much experience on that with the rest of the code base), so happy to change things to be more in line with sklearn development practices. 
I just wanted to try to give a best possible user experience, but sorry if that raised different discussions on changing standard practices :-)

amueller replied:
I think that's a good goal to have in mind :) And I'm not saying "no", I'm just saying if we want to change our practices we should give it due consideration.

jorisvandenbossche replied:
BTW, the reason the common tests were not failing on this, is because they are skipped for OneHotEncoder ... (https://github.com/scikit-learn/scikit-learn/blob/4b24fbedf5fa3b7b6b559141ad78708145b09704/sklearn/utils/testing.py#L553)

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+    # Deprecated keywords
+
+    @property
+    def n_values(self):
```
Doesn't the ``deprecate`` decorator do this deprecation warning? Also all warnings need version info.

jorisvandenbossche replied:
I am not fully sure `deprecated` decorator will work with a setter, but need to check (was not aware of this utility)

jorisvandenbossche replied:
Regarding the version info in the warnings: will add that (I wanted to ask if there are some contributor guidelines around what I need to put in it, but I found them myself :-) http://scikit-learn.org/stable/developers/contributing.html#deprecation)

amueller commented on sklearn/base.py:
```diff
@@ -225,12 +225,27 @@ def get_params(self, deep=True):
+            else:
+                param_names.append(p)
+
+        for key in param_names:
```
I would probably just catch deprecation warnings here.

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+from .label import LabelEncoder
+
+
+BOUNDS_THRESHOLD = 1e-7
```
Where is this used?

jorisvandenbossche replied:
sorry, that's a left-over from copying the code from `data.py`, not sure why my linter didn't catch that. Will remove.

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+]
+
+
+def _handle_zeros_in_scale(scale, copy=True):
```
where is that used?

jorisvandenbossche replied:
idem as above. Will remove

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+
+    def _handle_deprecations(self, X):
+
+        user_set_categories = False
```
I would remove this variable, it's used once and declaring it seems to make the code longer.

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+
+        # categories not set -> infer if we need legacy mode or not
+        elif self._deprecated_n_values != 'auto':
+            msg = (
```
Why is there two different messages in two different places for this deprecation?

jorisvandenbossche replied:
which other message do you mean? The one in the attribute? In that case: yes, I still need to update all actual message for the deprecated properties, as you also requested elsewhere (the current ones are just placeholders to already test the warnings)

amueller replied:
yes, I meant the ones in the attributes.

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,888 @@
+                    self._legacy_mode = True
+
+        # if user specified categorical_features -> always use legacy mode
+        if (not isinstance(self._deprecated_categorical_features,
```
I would move this logic up. I expect nearly all users of ``OneHotEncoder`` so far have specified this.
That is if we indeed don't want to support this going forward. As I argued, I would rather keep this feature.

jorisvandenbossche replied:
I case we keep the deprecation, what do you mean with "move this logic up"? Inside this function? (and in that case, what would be the purpose?)

amueller commented:
Sorry for flip-flopping on this but after heaving taught using CategoricalEncoder for a while, I really think keeping the ``categorical_features`` parameter would be really beneficial to our users.

jorisvandenbossche commented:
> I think your comment is cut off. 

Sorry, that was an editing left-over, the last sentence could be ignored (and removed it now)

> I think this might be a bit overkill. It's certainly not how we normally do things, it's more thorough. I'll go back to the code and check if I feel the thoroughness justifies the more complex code. We should also decide if this is what we want to do going forward. Because we might set a precedent and then always have to do it this way.

It is indeed more thorough than currently done, and @jnothman asked before also to propose this as a separate change (https://github.com/scikit-learn/scikit-learn/pull/10523#discussion_r163385500). The main problem is that I am not sure there would currently be use cases to apply this on in a possible PR. But maybe I can open a separate issue to propose this? (to at least keep the discussion separate)
Personally I think it is good to do thorough deprecation warnings, and it does not lead to too much complexity.

---

**On deprecating `categorical_features`**

@amueller:

> I'm somewhat tempted to not deprecate this and leave it in as a convenience. This is so common and would make sklearn so much more user friendly and not deprecating it won't kill us.
At least let's keep it until we have some other way to express #10603.

>  I expect nearly all users of OneHotEncoder so far have specified this.
That is if we indeed don't want to support this going forward. As I argued, I would rather keep this feature.

> Sorry for flip-flopping on this but after heaving taught using CategoricalEncoder for a while, I really think keeping the categorical_features parameter would be really beneficial to our users.

Some reasons to deprecate it:

- implementation wise: it would make the implementation more complex to keep supporting this (for now, I only kept this ability in the legacy mode, but did not expand the new code (copied from CategoricalEncoder as in master) to handle this). 
  The easiest would be to make this a `make_column_transformer((categorical_features, OneHotEncoder(**kwargs)))` under the hood, but that is of course not backwards compatible (eg for param names / attributes of the resulting estimator). But if we decide to keep this, it's certainly not hard to add custom logic for it in the new code as well.

- consistency wise: why should `OneHotEncoder` have a keyword to select the columns which to transform, and other transformers not? Certainly that now the answer for this use case is the ColumnTransformer. 
  That is also the reason I did not keep such a feature in the CategoricalEncoder API (and thus I also deprecated it here, in an effort to turn the existing OneHotEncoder into the new CategoricalEncoder API. But historical reasons to keep it are of course certainly valid).
  
But I certainly understand that having it there is a convenience feature, that people will have liked (I don't think that nearly all users of OneHotEncoder currently specify this, as I also have seen eg usecases where it is used after a LabelEncoder for some more manual preprocessing (so not in a pipeline)). 

But, if we really think `OneHotEncoder(categorical_features=[cat_columns])` is an important convenience over using `make_column_transformer(([cat_columns], OneHotEncoder()), remainder='passthrough')` (the `remainder='passthrough'` might not be needed if that becomes the default), I think we should discuss a more general solution. As this convenience would be useful for other transformers as well. 

For example, we could consider adding a `columns=` keyword to all relevant transformers that created the relevant ColumnTransformer under the hood (but that's maybe worth a separate issue)

amueller commented:
People are using OneHotEncoder after LabelEncoder outside a pipeline because we didn't support strings. I think once we support strings, people will want to simplify this horrible code that we currently force on them.

I think it's quite rare that there are only categorical columns, and so I think applying this to a subset of columns is an important case. That's not really true for other transformers (except the other categorical encoders). Because the categorical transformers are meant to work on a data representation that's different than the standard floating point representation that all of sklearn is working with.

In scikit-learn, entries of X are always float, except for in OneHotEncoder, Count/TfidfVectorizer and DictVectorizer. So I think having an exception for OneHotEncoder is not that bad.
The more general solution is certainly ColumnTransformer but making simple things simple is also important.

jorisvandenbossche commented:
It's true that wanting to apply it on a subset of the columns will occur more often for a OneHotEncoder or OrdinalEncoder, than for other transformers.

Just another idea: what would you think about keeping the functionality but deprecating the current implementation, where the new implementation would make use of the `ColumnTransformer` (`OneHotEncoder(categorical_features=..)` would be converted in a ColumnTransformer with a single transformer)?
Te advantage for me would be that it keeps the actual implementation consistent with the general solution, and having the keyword in the OneHotEncoder simply as a short-cut to create this general solution. 
The main disadvantage is that then accessing attributes on the returned object becomes more convoluted (something like `ohe.named_steps_['onehotencoder].categories_`)

jnothman reviewed (commented):
For parameter deprecations, I think we're best off just warning in `fit`. Yes, it's not perfect from a software engineering perspective, but it does okay and accords with our other conventions (validating in `fit`, etc.). I don't think showing the deprecated param in `__repr__` is the worst thing, but I can see why it adds noise.

As I suggested, I don't mind looking towards a more thorough approach to parameter deprecation in the future, but I don't think it needs to be innovated in this PR. I don't want deprecation process to hold up this PR, because as far as I'm concerned, if this is not in the next release, it's not going to be released.

jorisvandenbossche commented:
Yep, for the parameter deprecations, I was writing a new issue, so will leave that out of this PR and open a new one to discuss this separately.

amueller commented:
(sry misclick)

jorisvandenbossche committed:
Undo changes in base.py and deprecations of param attributes

jorisvandenbossche committed:
cleanup and updates

jorisvandenbossche committed:
update warning messages + use FutureWarning for integer case

jorisvandenbossche committed:
fix doctests

jorisvandenbossche committed:
also update deprecating messages in docstrings

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+    def __init__(self, n_values=None, categorical_features=None,
+                 categories=None, sparse=True, dtype=np.float64,
+                 handle_unknown='error'):
+        self._categories = self.categories = categories
```
I removed the deprecations of the params, and any validation in `fit`, but in the end I did keep the "underscored" versions of `categories`, `n_values` and `categorical_features`, to be able to know what the user passed itself, and thus raise the appropriate warning. 

The alternative is that throughout the actual implementation I have to add a lot of checks whether eg `self.categories is None` in addition to `self.categories == 'auto'` because None as default actually means 'auto' (and the same for the other two). 
But personally, I find that that makes the implementation much more complex compared to keeping those underscored versions of the initial passed values by the user (and setting the non-underscored versions to the appropriate default value in `_handle_deprecations` (where all that logic is then centralized).

jnothman replied:
But your underscores break set_params, which is unacceptable.

You can create all the underscore-prefixed attribs you like in fit to store validated versions and simplify downstream code. Your init must only set attributes by their arg name.

jorisvandenbossche replied:
Whoops, yes, that's a good catch. Indeed, I can create the underscored versions in `fit` and switch the purpose of the underscored vs non-underscored ones.

The problem I was trying to solve with keeping track of both, is that I wanted to warn on each fit, not only the first (because I translate certain values for `n_values` to `categories`, the second time you call `fit`, `categories` is set by the code, and no warning will be raised). But actually, maybe that's fine (or even better?) to only warn the first time?

jorisvandenbossche commented:
@amueller an additional complication regarding the question to keep `categorical_features` or to deprecate it, is the support for DataFrames.

Currently, it doesn't work for selecting columns based on the column names. It does work for integer indices or masks, simply because the dataframe is converted to an array in `check_array`.

```
In [1]: from sklearn.preprocessing import OneHotEncoder

In [4]: X = np.array([[0.1, 0.2, 0.3], [0, 1, 2]]).T

In [5]: X
Out[5]: 
array([[ 0.1,  0. ],
       [ 0.2,  1. ],
       [ 0.3,  2. ]])

In [6]: ohe = OneHotEncoder()

In [7]: ohe.fit_transform(X)
Out[7]: 
<3x4 sparse matrix of type '<class 'numpy.float64'>'
	with 6 stored elements in Compressed Sparse Row format>

In [8]: ohe.fit_transform(X).todense()  # <--- this is actually a strange behaviour for the floating point column
Out[8]: 
matrix([[ 1.,  1.,  0.,  0.],
        [ 1.,  0.,  1.,  0.],
        [ 1.,  0.,  0.,  1.]])

In [9]: ohe = OneHotEncoder(categorical_features=[1])

In [10]: ohe.fit_transform(X).todense()
Out[10]: 
matrix([[ 1. ,  0. ,  0. ,  0.1],
        [ 0. ,  1. ,  0. ,  0.2],
        [ 0. ,  0. ,  1. ,  0.3]])

In [13]: X_df = pd.DataFrame(X, columns=['a', 'b'])

In [14]: ohe.fit_transform(X_df).todense()
Out[14]: 
matrix([[ 1. ,  0. ,  0. ,  0.1],
        [ 0. ,  1. ,  0. ,  0.2],
        [ 0. ,  0. ,  1. ,  0.3]])

In [15]: ohe = OneHotEncoder(categorical_features=['b'])

In [16]: ohe.fit_transform(X_df).todense()
...
IndexError: arrays used as indices must be of integer (or boolean) type
```

So I think that *if* we keep it, it should have a full support (also dataframes) similarly to what ColumnTransformer uses. I suppose I will be able to reuse the helper functions from the ColumnTransformer module to deal with selecting columns, so it will not necessarily be harder compared to the current implementation. 
But given this limited dataframe support, I think a big chunk of users was not yet using the `categorical_featues` keyword (for whom a deprecation would not really be annoying).

Also, if we keep it in `OneHotEncoder`, we should probably also add it to the new OrdinalEncoder class for consistency within the encoders?

jorisvandenbossche commented:
@jnothman @amueller feedback on the updated version?

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+            warnings.warn(msg, DeprecationWarning)
+
+            # we internally translate this to the correct categories
+            # and don't use legacy mode
```
Does this complexity offer much benefit?

jorisvandenbossche replied:
I am not sure anymore why I did this in the beginning :-) 
I think it was to already use as much as possible the new code for most of the cases. But in the end, we want people to update their keywords (eg n_values -> categories), which then also makes sure they use the new code. 

So indeed, it's maybe a good idea to remove this part and just use legacy mode (the checks around this I want to keep to have good deprecation warnings).

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+
+            if isinstance(self.n_values, numbers.Integral):
+                n_features = X.shape[1]
+                self.categories = [
```
We conventionally avoid modifying constructor attributes in fit. Use private attributes if you need to...

jorisvandenbossche replied:
changed to use private attributes

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into categorical-refactor

jnothman reviewed (commented):
I have only looked at OHE implementation for now. Apart from concerns about set_params and underscored attributes, and maybe that the minimising of legacy mode is overkill, I think this design is pretty good. I'll need to find time to try break it 


In the new design it might make more sense to support `categories=[..., 'numeric',...]` to not encode some features than to support `categorical_features`

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+      categorical features.
+    sklearn.preprocessing.LabelEncoder : encodes target labels with values
+      between 0 and n_classes-1.
+    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
```
Not relevant here

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
+      dictionary items (also handles string-valued features).
+    sklearn.feature_extraction.FeatureHasher : performs an approximate one-hot
+      encoding of dictionary items or strings.
```
Not relevant here

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+        return X_int.astype(self.dtype, copy=False)
+
+    def inverse_transform(self, X):
+        """Convert back the data to the original representation.
```
Put "back" after "the data"

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+]
+
+
+def _transform_selected(X, transform, selected="all", copy=True):
```
Shouldn't this be shared with other preprocessing?

jorisvandenbossche replied:
No, OneHotEncoder is the only transformer that uses this (that has the ability to only transform a subset of the columns)

jnothman commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -0,0 +1,537 @@
+# Authors:
+#
+#          Giorgio Patrini
```
It's lonely up here

jorisvandenbossche replied:
This was just copied from `test_data.py` ... (so for sure he is also not the only author of that :-))
I quickly checked, and I don't think Giorgio contributed to the encoders tests, so is it fine to just remove this alltogether, or do you want me to add myself? (for the the actual implementation I did, not sure if it is needed for the tests)

jnothman commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -0,0 +1,537 @@
+# Make some data to be used many times
+rng = np.random.RandomState(0)
+n_features = 30
+n_samples = 1000
```
Not convinced there's any benefit to constants this big...

jorisvandenbossche replied:
whoops, more leftovers from copying it out of `test_data.py`, removing

jnothman reviewed (commented):
It would be easier to revue tests if you left them where they were! Would you mind noting on each test if it is:
* moved from CategoricalEncoder (i.e. testing new functionality)
* Only modified to hide warnings
* Otherwise
?

jnothman commented:
I don't think it's necessary to warn more than once

jorisvandenbossche committed:
clean init + don't modify constructor attributes in fit

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+    ----------
+    categories_ : list of arrays
+        The categories of each feature determined during fitting
+        (in order corresponding with output of ``transform``).
```
For now, this is always sorted (per feature), and it thus respects the order of the categories when specified as a param (since we for now require that they are sorted, as noted there). 

But I will add that they are in order of the features.

jorisvandenbossche commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,845 @@
+            msg = (
+                "Passing 'n_values' is deprecated in version 0.20 and will be "
+                "removed in 0.22. You can use the 'categories' keyword "
+                "instead. 'n_values=n' corresponds to 'categories=[range(n)]'."
```
Yep, good catch for the `n_features`. So that would translate to `categories=[range(n)] * n_features`

> Should we support the case where each feature has the same categories more explicitly, or is it too weird?

That would be nice. In principle we could check the first element of the list, and if that is not an array/list, assume it is single list/array of categories that is used for all features.

jorisvandenbossche reviewed (commented):
Thanks for the review!

I updated this to now have a fully clean `__init__` (no defintion of other variables) and to not modify constructor params in fit. 
This should fix the `set_params` problems (and added a test for that, as it wasn't catched)

> the minimising of legacy mode is overkill,

Yes, that is certainly something I don't feel strongly about, so can remove that (converting the `n_values` ourselves into the equivalent `categories` to use the new code)

> I think this design is pretty good.

It should be, you merged it for the CategoricalEncoder PR :-) 
Joking aside: certainly for the OrdinalEncoder, this is basically a copy of the design of the previously merged PR. But of course needs to be checked I didn't do anything stupid in splitting the class into two.

> It would be easier to revue tests if you left them where they were!

Yeah, I understand. I can certainly move them back if you want, but the `test_data.py` / `data.py` is getting long, so I thought it was a good opportunity to split. 

The moving is solely in the first commit, so you can also look at and comment on the diff without this commit (that should give a better idea of the actual changes): https://github.com/scikit-learn/scikit-learn/pull/10523/files/eb79ec939f965062072584fd000f7f07a465a912..90caab347b70f6a938225f65989dba20d65ac70f

But I will also go myself through the tests again to check more on this what actually changed. Eg for now I also left the test names as is (using "categorical_encoder") to reduce the diff, but in the end I should clean that up (use one_hot_encoder or ordinal_encoder where appropriate)

jorisvandenbossche committed:
cleanup for feedback Joel

jorisvandenbossche committed:
always use legacy mode when n_values is specified

jorisvandenbossche commented:
The last commit I added removed the complexity around minimising the use of the legacy mode: https://github.com/scikit-learn/scikit-learn/pull/10523/commits/a4eca9ecb1b44978caa862e46f4269a6648e080c. 
This indeed removed quite some complexity!

---

I think the main remaining discussion / decision point is what to do with `categorical_features`: 

1. deprecate
2. keep with limited functionality (no dataframe column names),
3. expand functionality (support same feature set as ColumnTransformer)

GaelVaroquaux commented:
> I think the main remaining discussion / decision point is what to do with categorical_features:

Does the original code supports dataframe in?

If not, I suggest that we deprecate.

jorisvandenbossche commented:
The answer is a bit double: it already supported dataframes when using integers (or boolean mask) to indicate columns, because the dataframe is simply converted to an array, and then it works exactly the same as the array input. 
But it does not support selecting columns by strings, as the ColumnTransformer does.

Therefore I think there are 3 options as I listed in the comment just above.

amueller commented:
```
In [8]: ohe.fit_transform(X).todense()  # <--- this is actually a strange behaviour for the floating point column
Out[8]: 
matrix([[ 1.,  1.,  0.,  0.],
        [ 1.,  0.,  1.,  0.],
        [ 1.,  0.,  0.,  1.]])
```
what is going on here?!

amueller commented:
> In the new design it might make more sense to support categories=[..., 'numeric',...] to not encode some features than to support categorical_features

pandas.get_dummies by default does everything that is object dtype and has the option to explicitly specify columns. I think we should be similarly flexible. You argued somewhere that you don't want your strings encoded by default. Do you think the situation is common where someone uses OneHotEncoder with a mix of integer, float and string values where they only want to encode the integers? That's kind of what you're proposing, right? Pandas would only one-hot encode the strings and not the integers by default. I think consistency with pandas by itself would be a benefit tbh.

jorisvandenbossche commented:
I think the default of `get_dummies` is interesting for OneHotEncoder if you use it as a standalone object, or as the only preprocessing step in a pipeline (without combination with ColumnTransformer). 
However, if you put it inside a ColumnTransformer, then that default would be a bit strange that not all the columns that you specified in the ColumnTransformer to one-hot-encode would actually be encoded (in case you have eg an integer column you want to one-hot encode). So that would be a bit annoying that you then still have to specify in the OneHotEncoder to encode all columns. 

So for me the ideal default differs between those two usecases, which is of course a bit difficult to implement :-)

GaelVaroquaux commented:
But the support for DataFrames is quite bad. We should push users to use
the ColumnTransformer. Hence, I think that we should deprecate selecting
columns.

amueller commented:
@GaelVaroquaux ``But the support for DataFrames is quite bad.`` what do you mean by that?

So you're ok with the simplest case of sklearn even without missing values being
```
make_pipeline(make_column_transformer(
    (OneHotEncoder(), categorical), (StandardScaler(), ~categorical),
   LinearRegression())
```
I have to say I'll be pretty embarrassed to present that as the first simple example of sklearn in any tutorial or workshop.

jorisvandenbossche commented:
> But the support for DataFrames is quite bad. what do you mean by that?

I think he means that the *current* support of OneHotEncoder for dataframes is quite bad, which it is. And it is an argument to say that not many dataframe-based users will currently use the `categorical_features` feature, and will be less impacted by deprecating it. 
But of course, because the *current* support is bad, it does not mean we cannot improve (reusing part of the code I wrote for the ColumnTransformer).

jorisvandenbossche commented:
> So you're ok with the simplest case of sklearn even without missing values being

I completely agree we should put more effort in finding ways to make simple things simple. But regarding the example you give, this won't improve with supporting `categorical_features` in OneHotEncoder, unless you are OK with also scaling the one hot encoded ones (which might be no problem). And once you also want to do imputation (and then you want different imputation for categorical vs numerical features), then you also need to use the ColumnTransformer, or you need a similar feature to distinguish the types of columns in the Imputer. 

I mainly want to say that we also can try to improve this basic usecase, without having this special case of selecting columns in OneHotEncoder (it is the only transformer that does it). 
For example, we could think about exposing the preprocessing part of the pipeline you show above as `SimplePreprocessor`.

amueller commented:
@jorisvandenbossche I'm still going back and forth between these... and yes, scaling one-hot-encoded variables is not super elegant but also probably not harmful. with the MinMaxScaler it wouldn't matter at all.
If we can actually agree on adding a ``SimplePreprocessor`` then I guess I'll be satisfied. I'm just afraid this will take another 3.5 years, which is how long it took us to find a consensus on the ColumnTransformer.

GaelVaroquaux commented:
> If we can actually agree on adding a SimplePreprocessor

Not too happy by the proposals that I've seen in this direction: the
default imputer that we have in not a good default choice (though we have
nothing clearly better today in the library).

jnothman commented:
Re SimplePreprocessor, start with an example then we can see where to go...

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -0,0 +1,827 @@
+
+    def _fit(self, X, handle_unknown='error'):
+
+        X_temp = check_array(X, dtype=None)
```
I know this is in CategoricalEncoder, but it might be a problem. Storing an array as strings can consume much memory.

jnothman replied:
Maybe these days lists of lists aren't so important. I'd say in the pre-pandas days they were. They still might be to some extent, e.g. from `csv.reader`.

jorisvandenbossche replied:
That's true, however fixing this is not that trivial (you need to "sniff" the data types in the full data to check if there are strings or not), so I would leave that for a separate issue if somebody is interested in it.

Also note that this will not slowdown existing code that dates from pre-pandas times, as before this was not possible at all with strings.

jnothman reviewed (commented):
I've not yet looked at those recent changes, but it would be good to get the documentation updates into the PR.

jorisvandenbossche committed:
Update documentation for split/rename

jorisvandenbossche committed:
fix doctests

jorisvandenbossche commented:
@jnothman included the documentation changes

jnothman commented on doc/modules/preprocessing.rst:
```diff
@@ -528,18 +526,19 @@ browsers was ordered arbitrarily).
 dummy encoding.
-This type of encoding is the default behaviour of the :class:`CategoricalEncoder`.
-The :class:`CategoricalEncoder` then transforms each categorical feature with
+This type of encoding can be obtained with the :class:`OneHotEncoderEncoder`,
```
EncoderEncoder?

jnothman commented on doc/modules/preprocessing.rst:
```diff
@@ -508,15 +508,13 @@ Such features can be efficiently coded as integers, for instance
 To convert categorical features to such integer codes, we can use the
-:class:`CategoricalEncoder`. When specifying that we want to perform an
-ordinal encoding, the estimator transforms each categorical feature to one
+:class:`OrdinalEncoder`.The estimator transforms each categorical feature to one
```
space after `.`.
The -> This

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -73,10 +73,10 @@ Classifiers and regressors
+  convert to ordinal integers.  Those two classes now handle
   encoding of all feature types (also handles string-valued features) and
   derives the categories based on the unique values in the features instead of
   the maximum value in the features. :issue:`9151` by :user:`Vighnesh Birodkar
```
please add this issue number

jnothman reviewed (commented):
WIP -> MRG?

jnothman reviewed (approved):
Given the amount of pre-existing content in the diff, I must say I don't have the energy to go through this line by line. I would've found it easier to review if everything stayed in data.py/test_data.py until a subsequent PR. However, I think this is looking good.

jorisvandenbossche committed:
doc fixes

jorisvandenbossche committed:
rename tests + remove one redundant one

jorisvandenbossche commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -0,0 +1,533 @@
+    return Xtr1.toarray()
+
+
+def test_categorical_encoder_onehot():
```
Updated the names

jorisvandenbossche commented:
> I would've found it easier to review if everything stayed in data.py/test_data.py until a subsequent PR.

@jnotham You can see (and comment on) the changes without the move in https://github.com/scikit-learn/scikit-learn/pull/10523/files/eb79ec939f965062072584fd000f7f07a465a912..419d2d9e0b741e72d321327f72a8b01bc8f9a4f1

That said, I am also fine with moving it back to `test_data.py`. However, should I then also move the code from `_encoders.py` back to `data.py`? 

(apart from doing it here or in a follow-up PR, I think the splitting is useful, as those data.py / test_data.py files were getting really long. But I didn't ask about this before starting this PR, so also fine with leaving it out from this PR and doing it in a follow-up)

Related to this, I think @amueller asked something about the import `from sklearn.preprocessing.data import OneHotEncoder` no longer working (but don't find the comment). 
So if we keep the splitting, I can re-import OneHotEncoder in `data.py` for backwards compatibility.

amueller commented:
travis failing ;)
I'll try to look at this "soon". (I just saw that I had replied I'm gonna do something "asap" in another place 6 month ago...)

jorisvandenbossche committed:
reword OneHotEncoder reference in the docs

jorisvandenbossche committed:
fix flake8

jorisvandenbossche committed:
fix doctest

jorisvandenbossche committed:
Suppress warning in RandomTreesEmbedding from using OneHotEncoder

jorisvandenbossche committed:
update 'see also' links to OneHotEncoder

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into categorical-refactor

jorisvandenbossche committed:
add similar dtype tests for new OneHotEncoder

jorisvandenbossche committed:
further fix doctest

jorisvandenbossche commented:
OK, made some small updates:

- some minor clean-up for pep8, doctests, correct cross-references (hopefully travis is happy now)
- suppressed warnings in `RandomTreesEmbedding` which was using a OneHotEncoder under the hood by adding a `categories='auto'` to that call (it is using integers, but the new behaviour should be identical in this case as the tree leave indices should be 0, 1, .. n)
- rebased for https://github.com/scikit-learn/scikit-learn/pull/11042 (that `dtype` is honoured) and added similar tests for the new OneHotEncoder (which already passed)

jnothman commented:
Great work!​

jorisvandenbossche committed:
fix doctest

jorisvandenbossche commented:
@jnothman @amueller can we try to move this forward to include in the release?

jnothman commented:
I'm happy with it as it is iirc

jorisvandenbossche committed:
suppress warnings again in the tests

ogrisel commented:
I am also ok with deprecating `categorical_features` and delegate feature dispatching to `ColumnTransformer`. We can always revise that decision later.

I would also introduce a stub `CategoricalEncoder` class that raises a TypeError in its constructor with explicit error message telling to use OneHotEncoder.

Even if  `CategoricalEncoder` was never part of a public scikit-learn release, I am affraid that is was already mentionned in several blog posts and even in a revised version of @ageron's book who we had the pleasure to chat with today. In @ageron's book there was a warning that it was an unreleased experimental feature but better be nice with the users and introduce that stub to help them upgrade their code quickly.

jnothman commented:
This PR doesn't yet remove CategoricalEncoder.

I'm okay with Olivier's suggestion:
```py
class CategoricalEncoder:
    "Removed"
    def __init__(*args, **kwargs):
        raise RuntimeError('CategoricalEncoder briefly existed in 0.19dev. '
                           'Its functionality has been rolled into '
                           'OneHotEncoder and OrdinalEncoder. This stub '
                           'will be removed in version 0.21.')
```

The only problem I see with it is that an ImportError would warn sooner.

jorisvandenbossche commented:
> This PR doesn't yet remove CategoricalEncoder.

It does, but will add the stub as proposed above.

> The only problem I see with it is that an ImportError would warn sooner.

Yeah, I was also thinking about that. But I don't see a way to both support `from sklearn.preprocessing import CategoricalEncoder` as `from sklearn import preprocessing; preprocessing.CategoricalEncoder()`. 
For the first I could add a `CategoricalEncoder.py` that raises an import error, but that doesn't work in the second case.

jorisvandenbossche committed:
add CategoricalEncoder stub with informative pointer

jnothman commented:
oh yeah... i always forget to load diff when looking for things

jorisvandenbossche commented:
@amueller are you OK with going merging this as is? (thus, deprecating `categorical_features` keyword)

jnothman commented:
I'm okay with that, yes.​

ogrisel reviewed (approved):
LGTM.

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2588,3 +2588,17 @@ def power_transform(X, method='box-cox', standardize=True, copy=True):
+
+class CategoricalEncoder:
+    """
+    CategoricalEncoder briefly existed in 0.19dev. Its functionality
```
sorry, should be 0.20dev

jnothman reviewed (commented):
Are we otherwise ready to give this a whirl?

jorisvandenbossche committed:
0.19 -> 0.20

TomDLT commented:
Great work !

nitpick: You may also want to update `examples/compose/column_transformer_mixed_types.py`

jnothman commented:
Great work, @jorisvandenbossche!

GaelVaroquaux commented:
Yesssss !

Congratulations team 

⁣Sent from my phone. Please forgive typos and briefness.​

On Jun 21, 2018, 12:14, at 12:14, Joel Nothman <notifications@github.com> wrote:
>Merged #10523.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/scikit-learn/scikit-learn/pull/10523#event-1693166398

ageron commented:
Awesome! Thanks everyone for your work on this important change, and special thanks to @jnothman 
and @ogrisel for adding the `CategoricalEncoder` stub, I really appreciate it.  :)

amueller commented:
yes, I'm ok with this for now ;) Thank you for all the work, this is great!
